server/src/main/java/org/elasticsearch/indices/recovery/RecoverySettings.java

```diff
@@ -434,4 +485,56 @@ Releasable tryAcquireSnapshotDownloadPermits() {
 
         return Releasables.releaseOnce(() -> maxSnapshotFileDownloadsPerNodeSemaphore.release(maxConcurrentSnapshotFileDownloads));
     }
+
+    private static ByteSizeValue recoveryMaxBytesPerSecBasedOnMemory(Settings settings) {
+        final boolean dedicatedColdOrFrozenNode = NODE_ROLES_SETTING.get(settings)
+            .stream()
+            .filter(DiscoveryNodeRole::canContainData)
+            .allMatch(dn -> dn.equals(DiscoveryNodeRole.DATA_COLD_NODE_ROLE) || dn.equals(DiscoveryNodeRole.DATA_FROZEN_NODE_ROLE));
+        if (dedicatedColdOrFrozenNode == false) {
+            // the node is not a dedicated cold and/or frozen node, use the default
+            return DEFAULT_MAX_BYTES_PER_SEC;
+        }
+        if (JavaVersion.current().compareTo(JavaVersion.parse("14")) < 0) {
+            // prior to JDK 14, the JDK did not take into consideration container memory limits when reporting total system memory
+            return DEFAULT_MAX_BYTES_PER_SEC;
+        }
+        /*
+         * Now we are looking at a node that is a dedicated cold and/or frozen node. In this case, we are going to set the recovery size as
+         * a function of the memory size. We are making an assumption here that the size of the instance is correlated with I/O resources.
+         * That is we are assuming that the larger the instance, the more disk and networking capacity it has available.
+         */
+        final ByteSizeValue totalPhysicalMemory = new ByteSizeValue(OsProbe.getInstance().getTotalPhysicalMemorySize());
+        final ByteSizeValue maxBytesPerSec;
+        if (totalPhysicalMemory.compareTo(new ByteSizeValue(4, ByteSizeUnit.GB)) <= 0) {
+            maxBytesPerSec = new ByteSizeValue(40, ByteSizeUnit.MB);
+        } else if (totalPhysicalMemory.compareTo(new ByteSizeValue(8, ByteSizeUnit.GB)) <= 0) {
+            maxBytesPerSec = new ByteSizeValue(60, ByteSizeUnit.MB);
+        } else if (totalPhysicalMemory.compareTo(new ByteSizeValue(16, ByteSizeUnit.GB)) <= 0) {
+            maxBytesPerSec = new ByteSizeValue(90, ByteSizeUnit.MB);
+        } else if (totalPhysicalMemory.compareTo(new ByteSizeValue(32, ByteSizeUnit.GB)) <= 0) {
+            maxBytesPerSec = new ByteSizeValue(125, ByteSizeUnit.MB);
+        } else {
+            maxBytesPerSec = new ByteSizeValue(250, ByteSizeUnit.MB);
+        }
+        return maxBytesPerSec;
+    }
+
+    private static ByteSizeValue recoveryMaxBytesPerSecBasedOnExternalSettings(Settings settings) {
+        assert NODE_ROLES_SETTING.get(settings).stream().anyMatch(DiscoveryNodeRole::canContainData);
+
+        final long diskBandwidth = NODE_DISK_AVAILABLE_BANDWIDTH_SETTING.get(settings).getBytes();
+        assert diskBandwidth > 0L : diskBandwidth;
```

Comment:
This assert does not match the minimum value in the setting, I wonder if we can add validation there to avoid it being set to 0? Or maybe for now just assert >= 0 here?